File: 2026美赛高频算法+创新算法汇总+源代码_副本/50多种常用算法源代码！（Matlab+Python）/Floyd 算法（全源最短路径算法）Matlab+Python代码/floyd_algorithm.m
% Floyd算法案例说明：
% Floyd算法（弗洛伊德算法）是一种用于寻找图中所有节点对之间最短路径的动态规划算法。
% 与Dijkstra算法不同，它可以一次性计算出所有节点之间的最短路径，适用于有向图和无向图，
% 但不适用于包含负权回路的图。
%
% 算法步骤：
% 1. 初始化距离矩阵dist，dist(i,j)表示节点i到节点j的初始距离
% 2. 对于每个中间节点k（1到n）：
%    a. 对于每对节点(i,j)：
%       i. 计算从i经过k到j的距离：dist(i,k) + dist(k,j)
%       ii. 如果该距离小于当前dist(i,j)，则更新dist(i,j)
% 3. 算法结束后，dist(i,j)即为节点i到节点j的最短距离
%
% 本案例使用一个包含5个节点的有向图，演示Floyd算法如何计算所有节点对之间的最短路径，
% 并以矩阵形式展示结果，同时可视化部分关键路径。

% 定义图的邻接矩阵（表示节点间的距离，Inf表示无直接连接）
% 这是一个有向图，节点1-5代表5个不同的地点
graph = [
    0   3   Inf 7   Inf;   % 节点1到其他节点的距离
    8   0   2   Inf Inf;   % 节点2到其他节点的距离
    Inf Inf 0   1   4;     % 节点3到其他节点的距离
    Inf Inf Inf 0   5;     % 节点4到其他节点的距离
    Inf 6   Inf Inf 0      % 节点5到其他节点的距离
];

% 节点名称（用于可视化）
node_names = {'A', 'B', 'C', 'D', 'E'};
num_nodes = size(graph, 1);

% 初始化距离矩阵
dist_matrix = graph;
% 初始化路径矩阵，path(i,j)表示i到j的最短路径中j的前一个节点
path_matrix = zeros(num_nodes, num_nodes);
path_matrix(:) = NaN;

% 初始化路径矩阵：如果i和j直接相连，则path(i,j) = i
for i = 1:num_nodes
    for j = 1:num_nodes
        if i ~= j && graph(i, j) > 0 && graph(i, j) < Inf
            path_matrix(i, j) = i;
        end
    end
end

% Floyd算法核心
fprintf('使用Floyd算法计算所有节点对之间的最短路径...\n');
for k = 1:num_nodes  % 中间节点
    for i = 1:num_nodes  % 起点
        for j = 1:num_nodes  % 终点
            % 如果i到k和k到j都可达，且经过k的路径更短
            if dist_matrix(i, k) + dist_matrix(k, j) < dist_matrix(i, j)
                % 更新最短距离
                dist_matrix(i, j) = dist_matrix(i, k) + dist_matrix(k, j);
                % 更新路径：j的前一个节点是k的前一个节点
                path_matrix(i, j) = path_matrix(k, j);
            end
        end
    end
end

% 输出所有节点对之间的最短距离矩阵
fprintf('\n所有节点对之间的最短距离矩阵：\n');
% 打印表头
fprintf('    ');
for j = 1:num_nodes
    fprintf('%4s', node_names{j});
end
fprintf('\n');

% 打印矩阵内容
for i = 1:num_nodes
    fprintf('%s: ', node_names{i});
    for j = 1:num_nodes
        if dist_matrix(i, j) >= Inf
            fprintf('%4s', '∞');
        else
            fprintf('%4d', dist_matrix(i, j));
        end
    end
    fprintf('\n');
end

% 根据路径矩阵获取从start到end的最短路径
function path = get_path(path_matrix, start, end_node)
    if isnan(path_matrix(start, end_node))
        path = [];  % 不可达
        return;
    end
    % 重建路径
    path = end_node;
    current = end_node;
    while current ~= start
        current = path_matrix(start, current);
        if isnan(current)  % 路径不存在
            path = [];
            return;
        end
        path = [current, path];
    end
end

% 输出部分关键路径
fprintf('\n部分关键路径详情：\n');
key_pairs = [1 5; 2 4; 5 3];  % 选择几对有代表性的节点（起点 终点）
for i = 1:size(key_pairs, 1)
    start = key_pairs(i, 1);
    end_node = key_pairs(i, 2);
    path = get_path(path_matrix, start, end_node);
    if isempty(path)
        fprintf('从%s到%s: 不可达\n', node_names{start}, node_names{end_node});
    else
        % 将节点索引转换为名称
        path_names = cell(1, length(path));
        for j = 1:length(path)
            path_names{j} = node_names{path(j)};
        end
        fprintf('从%s到%s: 路径: %s, 距离: %d\n', ...
            node_names{start}, node_names{end_node}, ...
            strjoin(path_names, ' -> '), dist_matrix(start, end_node));
    end
end

% 可视化图和部分最短路径
figure('Position', [100 100 800 600]);

% 定义节点位置（使用圆形布局）
theta = linspace(0, 2*pi, num_nodes+1);
theta = theta(1:end-1);  % 5个角度
radius = 2;
x = radius * cos(theta);
y = radius * sin(theta);

% 绘制所有节点和边
hold on;

% 绘制所有边（灰色）
for i = 1:num_nodes
    for j = 1:num_nodes
        if graph(i, j) > 0 && graph(i, j) < Inf
            % 绘制有向边
            arrow_x = [x(i), x(j)];
            arrow_y = [y(i), y(j)];
            plot(arrow_x, arrow_y, 'Color', [0.7 0.7 0.7], 'LineWidth', 1);
            % 添加箭头
            annotation('arrow', ...
                [arrow_x(1) arrow_x(2)]/max(abs(arrow_x))/2+0.5, ...
                [arrow_y(1) arrow_y(2)]/max(abs(arrow_y))/2+0.5, ...
                'Color', [0.7 0.7 0.7], 'LineWidth', 1, 'HeadWidth', 8);
        end
    end
end

% 绘制关键路径（红色高亮显示）
for i = 1:size(key_pairs, 1)
    start = key_pairs(i, 1);
    end_node = key_pairs(i, 2);
    path = get_path(path_matrix, start, end_node);
    if ~isempty(path) && length(path) > 1
        for j = 1:length(path)-1
            u = path(j);
            v = path(j+1);
            % 绘制有向边
            arrow_x = [x(u), x(v)];
            arrow_y = [y(u), y(v)];
            plot(arrow_x, arrow_y, 'r-', 'LineWidth', 2);
            % 添加箭头
            annotation('arrow', ...
                [arrow_x(1) arrow_x(2)]/max(abs(arrow_x))/2+0.5, ...
                [arrow_y(1) arrow_y(2)]/max(abs(arrow_y))/2+0.5, ...
                'Color', 'r', 'LineWidth', 2, 'HeadWidth', 8);
        end
    end
end

% 绘制节点
plot(x, y, 'bo', 'MarkerSize', 20, 'LineWidth', 2);

% 标注节点名称
for i = 1:num_nodes
    text(x(i)+0.1, y(i)+0.1, node_names{i}, 'FontSize', 12);
end

% 标注边权重
for i = 1:num_nodes
    for j = 1:num_nodes
        if graph(i, j) > 0 && graph(i, j) < Inf
            mid_x = (x(i) + x(j)) / 2;
            mid_y = (y(i) + y(j)) / 2;
            text(mid_x, mid_y, num2str(graph(i, j)), 'FontSize', 10, ...
                'BackgroundColor', 'w', 'EdgeColor', 'none');
        end
    end
end

title('Floyd算法计算的全源最短路径（红色为关键路径）', 'FontSize', 14);
axis equal;
axis off;
hold off;
    

File: 2026美赛高频算法+创新算法汇总+源代码_副本/50多种常用算法源代码！（Matlab+Python）/Floyd 算法（全源最短路径算法）Matlab+Python代码/floyd_algorithm.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Floyd算法案例说明：
Floyd算法（弗洛伊德算法）是一种用于寻找图中所有节点对之间最短路径的动态规划算法。
与Dijkstra算法不同，它可以一次性计算出所有节点之间的最短路径，适用于有向图和无向图，
但不适用于包含负权回路的图。

算法步骤：
1. 初始化距离矩阵dist，dist[i][j]表示节点i到节点j的初始距离
2. 对于每个中间节点k（0到n-1）：
   a. 对于每对节点(i, j)：
      i. 计算从i经过k到j的距离：dist[i][k] + dist[k][j]
      ii. 如果该距离小于当前dist[i][j]，则更新dist[i][j]
3. 算法结束后，dist[i][j]即为节点i到节点j的最短距离

本案例使用一个包含5个节点的有向图，演示Floyd算法如何计算所有节点对之间的最短路径，
并以矩阵形式展示结果，同时可视化部分关键路径。
"""

import numpy as np
import matplotlib.pyplot as plt
import networkx as nx

# 定义图的邻接矩阵（表示节点间的距离，inf表示无直接连接）
# 这是一个有向图，节点0-4代表5个不同的地点
graph = [
    [0, 3, np.inf, 7, np.inf],   # 节点0到其他节点的距离
    [8, 0, 2, np.inf, np.inf],   # 节点1到其他节点的距离
    [np.inf, np.inf, 0, 1, 4],   # 节点2到其他节点的距离
    [np.inf, np.inf, np.inf, 0, 5], # 节点3到其他节点的距离
    [np.inf, 6, np.inf, np.inf, 0]  # 节点4到其他节点的距离
]

# 节点名称（用于可视化）
node_names = ['A', 'B', 'C', 'D', 'E']
num_nodes = len(graph)

def floyd(graph):
    """实现Floyd算法，返回所有节点对之间的最短距离矩阵和路径矩阵"""
    # 初始化距离矩阵
    dist = [row[:] for row in graph]
    # 初始化路径矩阵，path[i][j]表示i到j的最短路径中j的前一个节点
    path = [[-1 for _ in range(num_nodes)] for _ in range(num_nodes)]
    
    # 初始化路径矩阵：如果i和j直接相连，则path[i][j] = i
    for i in range(num_nodes):
        for j in range(num_nodes):
            if i != j and dist[i][j] != np.inf:
                path[i][j] = i
    
    # Floyd算法核心
    for k in range(num_nodes):  # 中间节点
        for i in range(num_nodes):  # 起点
            for j in range(num_nodes):  # 终点
                # 如果i到k和k到j都可达，且经过k的路径更短
                if dist[i][k] + dist[k][j] < dist[i][j]:
                    # 更新最短距离
                    dist[i][j] = dist[i][k] + dist[k][j]
                    # 更新路径：j的前一个节点是k的前一个节点
                    path[i][j] = path[k][j]
    
    return dist, path

def get_path(path, start, end):
    """根据路径矩阵获取从start到end的最短路径"""
    if path[start][end] == -1:
        return None  # 不可达
    # 重建路径
    result = [end]
    current = end
    while current != start:
        current = path[start][current]
        if current == -1:  # 路径不存在
            return None
        result.append(current)
    # 反转路径，得到从起点到终点的顺序
    return result[::-1]

# 运行Floyd算法
print("使用Floyd算法计算所有节点对之间的最短路径...")
dist_matrix, path_matrix = floyd(graph)

# 输出所有节点对之间的最短距离矩阵
print("\n所有节点对之间的最短距离矩阵：")
# 打印表头
print("    " + " ".join(f"{name:4}" for name in node_names))
for i in range(num_nodes):
    row = [f"{node_names[i]}:"]
    for j in range(num_nodes):
        if dist_matrix[i][j] == np.inf:
            row.append("  ∞")
        else:
            row.append(f"{dist_matrix[i][j]:4}")
    print(" ".join(row))

# 输出部分关键路径
print("\n部分关键路径详情：")
key_pairs = [(0, 4), (1, 3), (4, 2)]  # 选择几对有代表性的节点
for start, end in key_pairs:
    path = get_path(path_matrix, start, end)
    if path is None:
        print(f"从{node_names[start]}到{node_names[end]}: 不可达")
    else:
        path_names = [node_names[node] for node in path]
        print(f"从{node_names[start]}到{node_names[end]}: "
              f"路径: {' -> '.join(path_names)}, 距离: {dist_matrix[start][end]}")

# 可视化图和部分最短路径
plt.figure(figsize=(12, 8))

# 创建有向图
G = nx.DiGraph()

# 添加节点
for i in range(num_nodes):
    G.add_node(i, name=node_names[i])

# 添加边和权重
for i in range(num_nodes):
    for j in range(num_nodes):
        if graph[i][j] > 0 and graph[i][j] != np.inf:
            G.add_edge(i, j, weight=int(graph[i][j]))

# 定义节点位置（使用圆形布局）
pos = nx.circular_layout(G)

# 绘制所有节点和边
nx.draw_networkx_nodes(G, pos, node_size=800, node_color='lightblue')
nx.draw_networkx_edges(G, pos, edgelist=G.edges(), width=1, 
                       alpha=0.5, edge_color='gray', arrowstyle='->')

# 绘制关键路径（用红色高亮显示）
for start, end in key_pairs:
    path = get_path(path_matrix, start, end)
    if path is not None and len(path) > 1:
        path_edges = [(path[i], path[i+1]) for i in range(len(path)-1)]
        nx.draw_networkx_edges(G, pos, edgelist=path_edges, width=3, 
                              edge_color='red', arrowstyle='->')

# 绘制节点标签和边权重
nx.draw_networkx_labels(G, pos, labels={i: node_names[i] for i in range(num_nodes)}, font_size=12)
edge_labels = {(i, j): graph[i][j] for i, j in G.edges()}
nx.draw_networkx_edge_labels(G, pos, edge_labels=edge_labels, font_size=10)

plt.title('Floyd算法计算的全源最短路径（红色为关键路径）', fontsize=14)
plt.axis('off')
plt.show()
    